File: swift/Sources/Wallet/CredentialsService.swift
import CredentialStore
import Foundation
import OpenID4VP
import SdJwt
import StatusList
import WalletAPI

/// Stored-credential management. Reads are local; `status` hits the network.
public struct CredentialsService {
    private let store: DefaultCredentialStore
    private let statusClient: StatusListClient

    init(store: DefaultCredentialStore, statusClient: StatusListClient) {
        self.store = store
        self.statusClient = statusClient
    }

    public func list(filter: CredentialFilter = .all) async throws -> [Credential] {
        try await store.list().map { $0.toCredential() }.filter { filter.matches($0) }
    }

    public func get(_ id: CredentialId) async throws -> Credential? {
        try await store.get(id)?.toCredential()
    }

    /// The raw serialized credential, for export / backup / inspection: the SD-JWT VC compact
    /// serialization for SD-JWT, or the base64url-encoded issuer-signed CBOR for mdoc. Nil if
    /// the credential is unknown or not yet issued.
    public func export(_ id: CredentialId) async throws -> String? {
        guard let envelope = try await store.get(id),
              case let .issued(_, instances) = envelope.lifecycle,
              let payload = instances.first?.payload else { return nil }
        switch envelope.format {
        case .sdJwtVc: return String(decoding: payload, as: UTF8.self)
        case .msoMdoc: return Base64Url.encode(payload)
        }
    }

    public func delete(_ id: CredentialId) async throws {
        try await store.delete(id)
    }

    /// Reactive list changes (added/updated/removed) for UI refresh.
    public func changes() async -> AsyncStream<CredentialChange> {
        let source = await store.changes()
        return AsyncStream { continuation in
            Task {
                for await change in source { continuation.yield(change.toFacade()) }
                continuation.finish()
            }
        }
    }

    /// Matches stored credentials against a DCQL query (OpenID4VP §6) — presentation-independent.
    public func match(_ dcqlJson: String) async throws -> CredentialMatch {
        let envelopes = try await store.list()
        let held = envelopes.compactMap { $0.toQueryable() }
        let query = try DcqlQuery.parse(JsonValue.parse(dcqlJson))
        let result = DcqlEngine.match(query, held: held)
        let byId = Dictionary(envelopes.map { ($0.id.value, $0) }, uniquingKeysWith: { first, _ in first })
        var byQuery: [String: [MatchedCredential]] = [:]
        for (queryId, candidates) in result.candidatesByQuery {
            byQuery[queryId] = candidates.compactMap { candidate in
                byId[candidate.credential.credentialId].map {
                    MatchedCredential(credential: $0.toCredential(), disclosedPaths: candidate.disclosedPaths)
                }
            }
        }
        return CredentialMatch(satisfiable: result.isSatisfiable(), byQuery: byQuery)
    }

    /// Revocation status via IETF Token Status List. Valid when the credential carries no status reference.
    public func status(_ id: CredentialId) async throws -> CredentialStatus {
        guard let claims = try await store.get(id)?.claimsTree() else { return .unknown }
        switch try await statusClient.check(claims: claims) {
        case .valid: return .valid
        case .invalid: return .invalid
        case .suspended: return .suspended
        case .unknown: return .unknown
        }
    }
}

/// Result of `CredentialsService.match`: which held credentials satisfy each query, and disclosed paths.
public struct CredentialMatch {
    public let satisfiable: Bool
    public let byQuery: [String: [MatchedCredential]]
}

public struct MatchedCredential {
    public let credential: Credential
    public let disclosedPaths: [[String]]
}

public enum CredentialChange {
    case added(CredentialId)
    case updated(CredentialId)
    case removed(CredentialId)
}

/// Credential revocation status (IETF Token Status List).
public enum CredentialStatus: Equatable { case valid, invalid, suspended, unknown }

extension CredentialStoreChange {
    func toFacade() -> CredentialChange {
        switch self {
        case let .added(id): return .added(id)
        case let .updated(id): return .updated(id)
        case let .removed(id): return .removed(id)
        }
    }
}

File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/CredentialsService.kt
package com.hopae.eudi.wallet

import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.spi.CredentialId
import com.hopae.eudi.wallet.status.StatusListClient
import com.hopae.eudi.wallet.store.CredentialStore
import com.hopae.eudi.wallet.store.CredentialStoreChange
import com.hopae.eudi.wallet.vp.DcqlEngine
import com.hopae.eudi.wallet.vp.DcqlQuery
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import com.hopae.eudi.wallet.status.CredentialStatus as StatusListStatus

/** Stored-credential management (API-CONTRACT.md §6.5). Reads are local; `status` hits the network. */
class CredentialsService internal constructor(
    private val store: CredentialStore,
    private val statusClient: StatusListClient,
) {

    suspend fun list(filter: CredentialFilter = CredentialFilter.All): List<Credential> =
        store.list().map { it.toCredential() }.filter { filter.matches(it) }

    suspend fun get(id: CredentialId): Credential? = store.get(id)?.toCredential()

    suspend fun delete(id: CredentialId) = store.delete(id)

    /** Reactive list changes (Added/Updated/Removed) for UI refresh. */
    val changes: Flow<CredentialChange> = store.changes.map { it.toCredentialChange() }

    /**
     * Revocation status via IETF Token Status List (network fetch, cached). Valid when the credential
     * carries no status reference. (SD-JWT VC; mdoc CWT status is a follow-up.)
     */
    suspend fun status(id: CredentialId): CredentialStatus {
        val claims = store.get(id)?.claimsTree() ?: return CredentialStatus.Unknown
        return when (statusClient.check(claims)) {
            StatusListStatus.VALID -> CredentialStatus.Valid
            StatusListStatus.INVALID -> CredentialStatus.Invalid
            StatusListStatus.SUSPENDED -> CredentialStatus.Suspended
            StatusListStatus.UNKNOWN -> CredentialStatus.Unknown
        }
    }

    /**
     * Matches stored credentials against a DCQL query (OpenID4VP §6) — presentation-independent.
     * Uses the same engine the presentation flow uses (credential_sets, claim_sets, null-wildcard).
     */
    suspend fun match(dcqlJson: String): CredentialMatch {
        val envelopes = store.list()
        val held = envelopes.mapNotNull { it.toQueryable() }
        val query = DcqlQuery.parse(JsonValue.parse(dcqlJson) as JsonValue.Obj)
        val result = DcqlEngine.match(query, held)
        val byId = envelopes.associateBy { it.id.value }
        return CredentialMatch(
            satisfiable = result.isSatisfiable(),
            byQuery = result.candidatesByQuery.mapValues { (_, candidates) ->
                candidates.mapNotNull { candidate ->
                    byId[candidate.credential.credentialId]?.let { MatchedCredential(it.toCredential(), candidate.disclosedPaths) }
                }
            },
        )
    }
}

/** Result of [CredentialsService.match]: which held credentials satisfy each query, and disclosed paths. */
class CredentialMatch(
    val satisfiable: Boolean,
    val byQuery: Map<String, List<MatchedCredential>>,
)

class MatchedCredential(val credential: Credential, val disclosedPaths: List<List<String>>)

/** Credential revocation status (IETF Token Status List). */
enum class CredentialStatus { Valid, Invalid, Suspended, Unknown }

sealed interface CredentialChange {
    val id: CredentialId

    data class Added(override val id: CredentialId) : CredentialChange
    data class Updated(override val id: CredentialId) : CredentialChange
    data class Removed(override val id: CredentialId) : CredentialChange
}

internal fun CredentialStoreChange.toCredentialChange(): CredentialChange = when (this) {
    is CredentialStoreChange.Added -> CredentialChange.Added(id)
    is CredentialStoreChange.Updated -> CredentialChange.Updated(id)
    is CredentialStoreChange.Removed -> CredentialChange.Removed(id)
}
